float64 s
float64 r
float64 h
